File: Framework/L2DMotionManager.cpp
/**
 *
 *  You can modify and use this source freely
 *  only for the development of application related Live2D.
 *
 *  (c) Live2D Inc. All rights reserved.
 */
#include "L2DMotionManager.h"

namespace live2d
{
	namespace framework
	{
		L2DMotionManager::L2DMotionManager()
		:currentPriority(0),reservePriority(0)
		{
			
		}
		
		
		int L2DMotionManager::startMotionPrio( AMotion* motion, bool isDelete, int priority )
		{
			if(priority==reservePriority)
			{
				reservePriority=0;
			}
			currentPriority=priority;
			return super::startMotion(motion, isDelete);
		}
		
		
		bool L2DMotionManager::updateParam(ALive2DModel* model)
		{
			bool updated=super::updateParam(model);
			if(isFinished())
			{
				currentPriority=0;
			}
			return updated;
		}
		
		
		bool L2DMotionManager::reserveMotion(int priority)
		{
			if ( priority <= reservePriority || priority <= currentPriority )
			{
				return false;
			}
			reservePriority=priority;
			return true;
		}
	}
}

File: package/live2d/v2/framework/motion/l2d_motion_manager.py
from ...core import MotionQueueManager


class L2DMotionManager(MotionQueueManager):

    def __init__(self):
        super().__init__()
        self.currentPriority = 0
        self.reservePriority = 0

    def getCurrentPriority(self):
        return self.currentPriority

    def getReservePriority(self):
        return self.reservePriority

    def reserveMotion(self, priority):
        if self.reservePriority >= priority:
            return False

        if self.currentPriority >= priority:
            return False

        self.reservePriority = priority
        return True

    def setReservePriority(self, val):
        self.reservePriority = val

    def updateParam(self, model):
        updated = super().updateParam(model)
        if self.isFinished():
            self.currentPriority = 0

        return updated

    def startMotionPrio(self, motion, priority):
        if priority == self.reservePriority:
            self.reservePriority = 0

        self.currentPriority = priority
        return self.startMotion(motion, False)
